pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1335,6 +1350,62 @@ public void deleteRealtimeTable(String tableName) {
   }
 
   /**
+   * Build the partition mapping table that maps a tuple of (partition number, replica group number) to a list of
+   * servers. Two important configurations are explained below.
+   *
+   * - 'numInstancesPerPartition': this number decides the number of servers within a replica group.
+   *
+   * - 'partitionColumn': this configuration decides whether to use the table or partition level replica groups.
+   *
+   * @param tableName: Name of table
+   * @param tableConfig: Configuration for table
+   * @return Partition mapping table from the given configuration
+   */
+  private PartitionToReplicaGroupMappingZKMetadata buildPartitionToReplicaGroupMapping(String tableName,
+      TableConfig tableConfig) {
+
+    // Fetch the server instances for the table.
+    List<String> servers = getServerInstancesForTable(tableName, TableType.OFFLINE);
+
+    // Fetch information required to build the mapping table from the table configuration.
+    String partitionColumn = tableConfig.getValidationConfig().getReplicaGroupStrategyConfig().getPartitionColumn();
+
+    // If we do not have the partition column configuration, we assume to use the table level replica groups,
+    // which is equivalent to have the same partition number for all segments (i.e. 1 partition).
+    int numPartitions = 1;
+    if (partitionColumn != null) {
+      numPartitions = tableConfig.getIndexingConfig().getSegmentPartitionConfig().getNumPartitions(partitionColumn);
+    }
+    int numReplicas = tableConfig.getValidationConfig().getReplicationNumber();
+    int numServers = servers.size();
+
+    ReplicaGroupStrategyConfig replicaGroupStrategyConfig = tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    int numInstancesPerPartition = replicaGroupStrategyConfig.getNumInstancesPerPartition();
+    int numServersPerReplicaGroup = numServers / numReplicas;
```

Comment:
We should not use 'numServersPerReplicaGroup' (basically this is the same as 'numInstancesPerPartition' right?)